src/main/java/jenkins/scm/impl/trait/WildcardSCMHeadFilterTrait.java

```diff
@@ -0,0 +1,144 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.scm.impl.trait;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import java.util.regex.Pattern;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.trait.SCMHeadPrefilter;
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceRequest;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTraitDescriptor;
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+/**
+ * Decorates a {@link SCMSource} with a {@link SCMHeadPrefilter} that filters {@link SCMHead} instances based on
+ * matching wildcard include/exclude rules.
+ *
+ * @since 2.2.0
+ */
+public class WildcardSCMHeadFilterTrait extends SCMSourceTrait {
+
+    /**
+     * The include rules.
+     */
+    @NonNull
+    private final String includes;
+
+    /**
+     * The exclude rules.
+     */
+    @NonNull
+    private final String excludes;
+
+    /**
+     * Stapler constructor.
+     *
+     * @param includes the include rules.
+     * @param excludes the exclude rules.
+     */
+    @DataBoundConstructor
+    public WildcardSCMHeadFilterTrait(@CheckForNull String includes, String excludes) {
+        this.includes = StringUtils.defaultIfBlank(includes, "*");
+        this.excludes = StringUtils.defaultIfBlank(excludes, "");
+    }
+
+    /**
+     * Returns the include rules.
+     *
+     * @return the include rules.
+     */
+    public String getIncludes() {
+        return includes;
+    }
+
+    /**
+     * Returns the exclude rules.
+     *
+     * @return the exclude rules.
+     */
+    public String getExcludes() {
+        return excludes;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected <B extends SCMSourceContext<B, R>, R extends SCMSourceRequest> void decorateContext(B context) {
+        context.withPrefilter(new SCMHeadPrefilter() {
+            @Override
+            public boolean isExcluded(@NonNull SCMSource request, @NonNull SCMHead head) {
+                return !Pattern.matches(getPattern(getIncludes()), head.getName())
+                        || (Pattern.matches(getPattern(getExcludes()), head.getName()));
+            }
+        });
+    }
+
+    /**
+     * Returns the pattern corresponding to the branches containing wildcards.
+     *
+     * @param branches the names of branches to create a pattern for
+     * @return pattern corresponding to the branches containing wildcards
+     */
+    private String getPattern(String branches) {
+        StringBuilder quotedBranches = new StringBuilder();
+        for (String wildcard : branches.split(" ")) {
```

Comment:
Maybe this is copied from somewhere to keep bahaviour from previous versions, otherwise there are nice utility methods in ANT that enables `**` and `?` style wildcards as well.